internal/builders/docker/pkg/builder.go

```diff
@@ -476,34 +478,74 @@ func (c *GitClient) checkoutGitCommit() error {
 	return nil
 }
 
+type tempFileResult struct {
+	File *os.File
+	Err  error
+}
+
+// A helper function used by saveToTempFile to process one individual file.
+func saveOneTempFile(verbose bool, reader io.Reader, fileChannel chan tempFileResult, printChannel chan string) {
+	var allBytes []byte
+	scanner := bufio.NewScanner(reader)
+	for scanner.Scan() {
+		bytes := scanner.Bytes()
+		allBytes = append(allBytes, bytes...)
+		allBytes = append(allBytes, '\n')
+
+		if verbose {
+			printChannel <- string(bytes)
+		}
+	}
+
+	tmpfile, err := os.CreateTemp("", "log-*.txt")
+	if err != nil {
+		fileChannel <- tempFileResult{Err: err}
+		return
+	}
+	if _, err := tmpfile.Write(allBytes); err != nil {
+		tmpfile.Close()
+		fileChannel <- tempFileResult{Err: fmt.Errorf("couldn't write bytes to tempfile: %v", err)}
+	}
+
+	fileChannel <- tempFileResult{File: tmpfile}
+}
+
 // saveToTempFile creates a tempfile in `/tmp` and writes the content of the
 // given readers to that file.
+// It processes all provided readers concurrently.
 func saveToTempFile(verbose bool, readers ...io.Reader) ([]string, error) {
-	var files []string
+	if verbose {
+		fmt.Print("\n\n>>>>>>>>>>>>>> output from command <<<<<<<<<<<<<<\n")
+	}
+	var wg sync.WaitGroup
+	var fileChannel = make(chan tempFileResult, len(readers))
+	var printChannel = make(chan string)
+
 	for _, reader := range readers {
-		bytes, err := io.ReadAll(reader)
-		if err != nil {
-			return files, err
-		}
+		wg.Add(1)
+		go func(reader io.Reader) {
+			defer wg.Done()
+			saveOneTempFile(verbose, reader, fileChannel, printChannel)
+		}(reader)
+	}
 
-		if verbose {
-			if len(bytes) > 0 {
-				fmt.Print("\n\n>>>>>>>>>>>>>> output from command <<<<<<<<<<<<<<\n")
-				fmt.Printf("%s", bytes)
-				fmt.Print("=================================================\n\n\n")
-			}
-		}
+	// Close the channel once all goroutines have finished.
+	go func() {
```

Comment:
I think this should not be in a go routine?

Reply:
Either this needs to be in a goroutine, of lines 539-541 below need to be a in a goroutine, so that printing can occur while the reader processing goroutines are running.

Reply:
ho I see it now. That makes sense. Maybe add a comment

Reply:
I added a few other comments as well. Let me know if it's too much, or if you have other questions that we can capture in some comments here.

Reply:
Wouldn't it also be okay to just move this at the end of the parent function?

Reply:
We need this to be running concurrently before we range over the channels below, because those range loops are terminated by the channel closing.

Reply:
ah yes. My idea would've been to use the same routine, but only start waiting after those loops conclude. But you're right, in this case the loops would never conclude. Thanks for explaining.